src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -263,18 +265,25 @@ bool ZeroMQBackend::DoPublishEvent(const std::string& topic, const std::string&
     return true;
 }
 
-bool ZeroMQBackend::DoSubscribe(const std::string& topic_prefix) {
+bool ZeroMQBackend::DoSubscribe(const std::string& topic_prefix, SubscribeCallback cb) {
     ZEROMQ_DEBUG("Subscribing to %s", topic_prefix.c_str());
     try {
         // Prepend 0x01 byte to indicate subscription to XSUB socket
         // This is the XSUB API instead of setsockopt(ZMQ_SUBSCRIBE).
         std::string msg = "\x01" + topic_prefix;
-        xsub.send(zmq::const_buffer(msg.data(), msg.size()));
+        main_inproc.send(zmq::const_buffer(msg.data(), msg.size()));
     } catch ( zmq::error_t& err ) {
         zeek::reporter->Error("Failed to subscribe to topic %s: %s", topic_prefix.c_str(), err.what());
+        if ( cb )
+            cb(topic_prefix, {CallbackStatus::Error, err.what()});
+
         return false;
     }
 
+    // Store the callback for later.
+    if ( cb )
+        subscription_callbacks[topic_prefix] = cb;
```

Comment:
Nit: we can avoid creating a default entry altogether with

```suggestion
        subscription_callbacks.insert({topic_prefix, cb});
```